staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset_pruner.go

```diff
@@ -19,15 +19,128 @@ package apply
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/client-go/dynamic"
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
 )
 
 type applySetPruner struct {
+	dynamicClient dynamic.Interface
+
+	visitedUids sets.Set[types.UID]
+
+	cascadingStrategy metav1.DeletionPropagation
+	dryRunStrategy    cmdutil.DryRunStrategy
+	gracePeriod       int
+
+	printer printers.ResourcePrinter
+
+	out io.Writer
+}
+
+func newApplySetPruner(o *ApplyOptions) (*applySetPruner, error) {
+	printer, err := o.ToPrinter("pruned")
+	if err != nil {
+		return nil, err
+	}
+
+	return &applySetPruner{
+		dynamicClient: o.DynamicClient,
+
+		visitedUids: o.VisitedUids,
+
+		cascadingStrategy: o.DeleteOptions.CascadingStrategy,
+		dryRunStrategy:    o.DryRunStrategy,
+		gracePeriod:       o.DeleteOptions.GracePeriod,
+
+		printer: printer,
+
+		out: o.Out,
+	}, nil
+}
+
+func formatGVKForHumans(gvk schema.GroupVersionKind) string {
+	gk := gvk.GroupKind()
+	return gk.Kind
+}
+
+func formatNamespacedNameForHumans(id types.NamespacedName) string {
+	if id.Namespace == "" {
+		return id.Name
+	}
+	return id.Namespace + "/" + id.Name
 }
 
-func newApplySetPruner(_ *ApplyOptions) *applySetPruner {
-	return &applySetPruner{}
+func (p *applySetPruner) pruneAll(ctx context.Context, applyset *ApplySet) error {
+	// TODO: Split into discovery and deletion, run discovery in parallel (and maybe in consistent order or in parallel?)
+	for _, restMapping := range applyset.AllPrunableResources() {
+		switch restMapping.Scope.Name() {
+		case meta.RESTScopeNameNamespace:
+			for _, namespace := range applyset.AllPrunableNamespaces() {
+				if namespace == "" {
+					// Just double-check because otherwise we get cryptic error messages
+					return fmt.Errorf("unexpectedly encountered empty namespace during prune of namespace-scoped resource %v", restMapping.GroupVersionKind)
+				}
+				if err := p.prune(ctx, namespace, restMapping, applyset); err != nil {
+					return fmt.Errorf("pruning %v objects: %w", formatGVKForHumans(restMapping.GroupVersionKind), err)
+				}
+			}
+
+		case meta.RESTScopeNameRoot:
+			if err := p.prune(ctx, metav1.NamespaceNone, restMapping, applyset); err != nil {
+				return fmt.Errorf("pruning %v objects: %w", formatGVKForHumans(restMapping.GroupVersionKind), err)
+			}
+
+		default:
+			return fmt.Errorf("unhandled scope %q", restMapping.Scope.Name())
+		}
+	}
+
+	return nil
+}
+
+func (p *applySetPruner) prune(ctx context.Context, namespace string, mapping *meta.RESTMapping, applyset *ApplySet) error {
+	applysetLabelSelector := applyset.LabelSelectorForMembers()
+
+	opt := metav1.ListOptions{
+		LabelSelector: applysetLabelSelector,
+	}
+
+	objects, err := p.dynamicClient.Resource(mapping.Resource).Namespace(namespace).List(ctx, opt)
+	if err != nil {
+		return err
+	}
+
+	for i := range objects.Items {
+		obj := &objects.Items[i]
+
+		uid := obj.GetUID()
+		if p.visitedUids.Has(uid) {
+			continue
+		}
+		name := obj.GetName()
+		id := types.NamespacedName{
```

Comment:
So in general I try not to pass multiple arguments of the same type because it's too easy to transpose arguments and then you don't get any help from the compiler.  But in this case it seems a relatively marginal win, so fixed!